Merge Improver and TerminalGrounder classes into master of original repo

RedFantom opened the pull request:
These commits contain the added `Improver` and `TerminalGrounder` classes. The `Improver` class should allow more images to be segmented because it allows the modification of color saturation, sharpness, brightness and contrast with default values that helps to improve most images. This is to make the segmentation of images easier. Also added is the `TerminalGrounder` class, which allows the grounding of image files with an interactive shell. The image cannot be shown in the terminal, but the amount of segments is provided. The changes in this branch depend on `PIL`.

Note: the images in this branch were supposed to be used for additional testing of the `Improver` class, however, this is not yet implemented. Should they be kept or should they be kicked? @goncalopp 

This PR is related to issue #14 . Squash merging this branch with a new commit message is recommended.

goncalopp commented:
Yes, the images and `travis.yml` should be excluded for now.

What's the intended use for `TerminalGrounder`? Why not just use `TextGrounder` if you're on the terminal?

RedFantom commented:
Well, if you know how many characters your image is supposed to have and you know that it is the same as the amount of segments found, you can also ground it interactively in the Terminal. I admit that the use-case is probably limited to only a very small percentage, but I thought it should be an option.

RedFantom commented:
I'll try rebasing these commits soon. I hope to have time for it upcoming Friday, maybe Thursday, but certainly not earlier.

goncalopp commented:
Should we do this one first? Or the unit tests?

RedFantom commented:
This branch is now rebased successfully onto `master`, and in the process I cleaned up some more things that I apparently missed before, sorry :( . There are no tests available for them at the moment though, because the images that I added to test segmentation have been removed. I could re-add them and then add the unittests again though.

RedFantom commented on improver.py:
```diff
@@ -0,0 +1,75 @@
+    """
+    Improver class that provides a wrap around the ImageEnhance classes from PIL/pillow
+    The Improved image can be accessed through object.image, this is a PIL/pillow RGB
+    format image, that is NOT compatible with the cv2 based files.ImageFile class!
```
Yes, that is the case. After the improvements have been applied, the image in the `ImageFile` object is altered so the improved image can be used with the OCR functions without altering the original file. It would be possible to allow the user to put in Pillow image files as well by doing instance checking, but I don't know whether that would actually make things easier as you then support a data format *as input* that isn't used in the whole rest of the code.

goncalopp replied:
What I meant is that the internal state (`self.image`) of ImageImprover is a PIL image.

It seems like this (`ImageImprover`) could more easily be done by two standalone functions (`enhance`, `crop`) that take the image as the argument and could then be used directly from `ImageFileImprover`. That would avoid the code duplication as well, and you wouldn't necessarily need to keep a PIL image at all, since you can always convert the input image from cv to PIL, and convert the output from PIL to cv before returning. It's probably not very efficient, but avoids ever exposing PIL images - what do you think? Would this work with your fork code so far?

RedFantom replied:
That is probably a good idea. That would work fine with the code, and you can convert the image back to the `cv2` format upon returning it. I will change the class in this branch to a set of functions in that case.

RedFantom commented on opencv_utils.py:
```diff
@@ -94,7 +94,7 @@ def background_color(image, numpy_result=True):
 def show_image_and_wait_for_key(image, name="Image"):
     """
     Shows an image, outputting name. keygroups is a dictionary of keycodes to functions;
-    they are executed when the corresponding keycode is pressed:return:
```
Actually I think it was me, but it needs fixing, in any case ;) .

RedFantom commented:
Sorry for putting the `TerminalGrounder` in the list on the [wiki page](https://github.com/goncalopp/simple-ocr-opencv/wiki/Grounding) already, I was a bit overenthusiastic -_-'.

goncalopp commented:
Thanks, this is looking much better :)

I didn't fully understand why you added individual functions though, for what was a parameter before? 
I changed it a bit to work with default arguments, can you please take a look  on 8362ef97605110ad65b4f143539357ecfcf17486 (untested) and let me know what you think?

>Sorry for putting the TerminalGrounder in the list on the wiki page already, I was a bit overenthusiastic -_-'.

No worries at all, thanks for improving documentation!

goncalopp commented:
I think the repetition of the `image = ImageEnhance.X(image).enhance(X)` could be improved by taking arguments as `kwargs`, but it would hurt readability a lot, so it's probably not worth it for reducing 5 lines or so

RedFantom commented:
Perhaps making the default argument `None` could work as well, but I mainly added the individual functions so that the improvements can also be executed by the user in a different order than the one pre-programmed in `enhance_image`, because the order in which the functions are executed can have significant influence on the end result and I thought it may be best to allow the user to execute the functions in a custom order as well. I know it's not the most efficient way of implementing that, as in every step the user does two conversions that are not immediately required (to and from Pillow format). What would of course be possible is writing one function that takes a `list` of classes as an argument and loops through that, but I don't know whether that would be very readable. Perhaps if I add it to the documentation as well (I was planning to) that it would be feasible?

RedFantom commented:
Yes, that would be a perfectly viable option now, using the keyword arguments. I just thought it might not be clear enough, but I guess it is then! In that case I will cherry-pick your commit and push it here!

goncalopp commented on grounding.py:
```diff
@@ -39,6 +68,7 @@ def ground(self, imagefile, segments, _=None):
             classes = classes_from_numpy(imagefile.ground.classes)
             segments = imagefile.ground.segments
         else:
+            # char(10) is newline. it represents a non-assigned label, and will b filtered
```
I don't this should be here? This is a outdated comment

RedFantom replied:
Oh, OK. I wasn't sure, so I left it in when rebasing. I'll  take it out ASAP.

goncalopp commented on tests/test_grounding.py:
```diff
@@ -43,3 +41,24 @@ def mock_input(*args):
+        segmenter = ContourSegmenter()
+        image = ImageFile('digits1')
+        segments = segmenter.process(image.image)
+        characters = ['9', '8', '2', '1', '4', '8', '0', '8', '6', '5', '1', '3', '2', '8',
```
We don't need to be hardcoding/duplicating the character sequence over and over. You can just use the already provided ground data - see how it's working in the previous test

goncalopp commented:
Alright, this is looking pretty good, thanks! :)
I did a final pass over the code, there's only two minor issues

RedFantom commented:
The tests pass on my Windows machine with OpenCV 3 as well.

RedFantom commented:
Looks good! My commit earlier was rushed as well, so I actually just thought of it last night that I should've removed more code repetition.
